File: DESIGN-MOD/prototype.js
//心形原型
function HeartShape(color, size){
    this.color = color;
    this.size = size;
}

HeartShape.prototype.clone = function(){
   return new HeartShape(this.color, this.size);
};

//心形工厂
function HeartShapeFacotry(obj){
    this.obj = obj;
}
HeartShapeFacotry.prototype.getHeartShape = function(){
    return this.obj.clone();
}

var heartShapeFacotry = new HeartShapeFacotry(new HeartShape('red', 'big'));

//红色大心形，要多少有多少
var a = heartShapeFacotry.getHeartShape();
var b = heartShapeFacotry.getHeartShape();

console.log(a.color)
console.log(b.color)

console.log('--------------')

a.color = 'yellow';
console.log(a.color)
console.log(b.color)

/*print out

red
red
--------------
yellow
red
*/

File: DESIGN-MOD/prototype.py
#-*- coding:utf-8 -*-

from copy import deepcopy

#心形原型
class HeartShape(object):
    def __init__(self, color, size):
        self.color = color
        self.size = size
    def clone(self):
        return deepcopy(self)

#心形工厂
class HeartShapeFactory(object):
    def __init__(self, obj):
        self.obj = obj
    def getHeartShape(self):
        return self.obj.clone()

if __name__ == "__main__":
    heartShapeFactory = HeartShapeFactory(HeartShape("red", "big"))

    #红色大心形，要多少有多少
    a = heartShapeFactory.getHeartShape()
    b = heartShapeFactory.getHeartShape()

    print a.color
    print b.color

    print "------------------"
    
    a.color = "yellow"
    print a.color
    print b.color

"""print out

red
red
------------------
yellow
red
"""